File: 백준/Gold/1722. 순열의 순서/순열의 순서.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        boolean[] used = new boolean[n + 1];
        long[] factorial = new long[n + 1];
        factorial[0] = 1;
        for (int i = 1; i <= n; i++) {
            factorial[i] = factorial[i - 1] * i;
        }

        StringTokenizer st = new StringTokenizer(br.readLine());
        int command = Integer.parseInt(st.nextToken());

        StringBuilder sb = new StringBuilder();
        if (command == 1) {
            long k = Long.parseLong(st.nextToken());
            int[] result = new int[n + 1];

            for (int i = 1; i <= n; i++) {
                int count = 1;

                for (int j = 1; j <= n; j++) {
                    if (used[j]) continue;

                    if (k <= count * factorial[n - i]) {
                        k -= (count - 1) * factorial[n - i];
                        result[i] = j;
                        used[j] = true;
                        break;
                    }

                    count++;
                }
            }

            for (int i = 1; i <= n; i++) {
                sb.append(result[i]).append(" ");
            }
        } else {
            int[] permutation = new int[n + 1];
            for (int i = 1; i <= n; i++) {
                permutation[i] = Integer.parseInt(st.nextToken());
            }

            long order = 1;

            for (int i = 1; i <= n; i++) {
                long smallerCount = 0;

                for (int j = 1; j < permutation[i]; j++) {
                    if (!used[j]) smallerCount++;
                }

                order += smallerCount * factorial[n - i];
                used[permutation[i]] = true;
            }

            sb.append(order);
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/1722. 순열의 순서/순열의 순서.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()

    val used = BooleanArray(n + 1)
    val factorial = LongArray(n + 1)
    factorial[0] = 1
    for (i in 1..n) {
        factorial[i] = factorial[i - 1] * i
    }

    val input = br.readLine().split(" ")
    val command = input[0].toInt()

    val sb = StringBuilder()
    if (command == 1) {
        var k = input[1].toLong()
        val result = IntArray(n + 1)

        for (i in 1..n) {
            var count = 1

            for (j in 1..n) {
                if (used[j]) continue

                if (k <= count * factorial[n - i]) {
                    k -= (count - 1) * factorial[n - i]
                    result[i] = j
                    used[j] = true
                    break
                }

                count++
            }
        }

        sb.append(result.slice(1..n).joinToString(" "))
    } else {
        val permutation = input.drop(1).map { it.toInt() }
        var order = 1L

        for (i in 0 until n) {
            var smallerCount = 0L

            for (j in 1 until permutation[i]) {
                if (!used[j]) smallerCount++
            }

            order += smallerCount * factorial[n - i - 1]
            used[permutation[i]] = true
        }

        sb.append(order)
    }

    print(sb)
}
